File: flutter-application-main/lib/models/rally/stage_activity.dart
import 'package:dart_mappable/dart_mappable.dart';

import 'stage_activity_answer.dart';

@MappableClass(discriminatorKey: "type")
abstract class StageActivity {
  final String id;
  final String title;
  final String task;
  final int maxPoints;

  StageActivity({
    required this.id,
    required this.title,
    required this.task,
    required this.maxPoints,
  });
}

abstract class ClosedQuestionActivity extends StageActivity {
  final List<StageActivityAnswer> answers;

  ClosedQuestionActivity({
    required String id,
    required String title,
    required String task,
    required int maxPoints,
    required this.answers,
  }) : super(id: id, title: title, task: task, maxPoints: maxPoints);
}

@MappableClass(discriminatorValue: "singleChoice")
class SingleChoiceActivity extends ClosedQuestionActivity {
  SingleChoiceActivity({
    required String id,
    required String title,
    required String task,
    required int maxPoints,
    required List<StageActivityAnswer> answers,
  }) : super(
          id: id,
          title: title,
          task: task,
          maxPoints: maxPoints,
          answers: answers,
        );
}

@MappableClass(discriminatorValue: "multipleChoice")
class MultipleChoiceActivity extends ClosedQuestionActivity {
  MultipleChoiceActivity({
    required String id,
    required String title,
    required String task,
    required int maxPoints,
    required List<StageActivityAnswer> answers,
  }) : super(
          id: id,
          title: title,
          task: task,
          maxPoints: maxPoints,
          answers: answers,
        );
}

@MappableClass(discriminatorValue: "textInput")
class TextInputActivity extends StageActivity {
  TextInputActivity(
      {required String id,
      required String title,
      required String task,
      required int maxPoints})
      : super(id: id, title: title, task: task, maxPoints: maxPoints);
}

File: frontend-main/src/models/rally/stageActivity.ts
export enum StageActivityType {
  singleChoice = <any>"singleChoice",
  multipleChoice = <any>"multipleChoice",
  textInput = <any>"textInput",
}

export class StageActivity {
  id: string;
  title: string;
  task: string;
  maxPoints: number;
  type: StageActivityType;
  answers: StageActivityAnswer[];

  constructor(
    id: string,
    title: string,
    task: string,
    maxPoints: number,
    type: StageActivityType,
    answers: StageActivityAnswer[]
  ) {
    this.id = id;
    this.title = title;
    this.task = task;
    this.maxPoints = maxPoints;
    this.type = type;
    this.answers = answers;
  }
}

export class StageActivityAnswer {
  id: string;
  answerText: string;
  isCorrect: boolean;

  constructor(id: string, answerText: string, isCorrect: boolean) {
    this.id = id;
    this.answerText = answerText;
    this.isCorrect = isCorrect;
  }
}
